### PR TITLE
Autogen build deps, allow building on macos

### PR DESCRIPTION
**Pull request by wez**
There are a couple of commits in this PR:

* `build.rs` change to restrict the header files that `bindgen` can see to those of the target compiler, rather than pulling in headers from the host compiler.
* `build.rs` change to find the sources and headers at build time and emit the appropriate deps for cargo.  Removing the false dep on `memory.x` speeds up the build too.  This change means that we compile all of the code in the bundled SDK and some of it is missing pieces (ANT and RTOS header files), so the most expedient resolution is just to delete the source files that want those(!)
* I ran `cargo fmt` over the code

Some folks would prefer that these be separate PRs, but before I go through the effort of submitting them independently I wanted to reach out and see what you think.

### CODE REVIEW

**Comment by jamesmunns** [comment]
Hey! Thanks for the contributions, It'll take me a little bit to grok all the changes. Without looking at the code too deeply, I'd like to make sure the following things are true:

- Any changes to the build behavior should work on Linux or OSX. You can definitely use the provided dockerfile to test this if you don't have a linux machine handy
- I'm not entirely against adding Python as a dependency, but it would be nice if you could do it in Rust, so we don't increase the number of dependencies.
- Please update any docs with regards to required dependencies or setup steps. I would greatly appreciate an install guide for OSX.
- It looks like you are using Rust's `feature` system to generate settings, you might want to be careful with this, as this operates as a union (e.g. if multiple co/sub dependencies request different features, cargo will take the union of all requested features). I don't think this is a problem here, but you might know better :) 

Thanks again for your help, I'll try to test your changes locally in the next day or two, as well as look a little bit deeper. I definitely appreciate the contribution!

File: build.rs [hunk 1]
```diff
@@ -8,7 +8,7 @@ use std::process::Command;
 use std::fs::File;
 use std::io::Write;
 
-use gcc::Config;
+use gcc::Build;
```
Comment (anxiousmodernman) [inline build.rs]: We might need to this (and maybe other build.rs functions) behind a macos platform flag. I tested this branch with the following:

```
git remote add wez https://github.com/wez/nrf52dk-sys
git fetch wez
git checkout wez/master
rm -rf target
xargo build --release
```

I get this error on linux

wez_master

Comment (anxiousmodernman) [inline build.rs:12]: Hmmm, okay I take that back. I tried updating gcc and I no longer am missing `Build`, but I get other errors.

upgraded_gcc

Something funky in the generated code. Using these versions
`bindgen 0.31.3`
`xargo 0.3.8`

File: build.rs [hunk 2]
```diff
@@ -294,6 +311,11 @@ static FLAGS: &[&str] = &["-std=c99",
                           "-fdata-sections",
                           "-fno-strict-aliasing",
                           "-fno-builtin",
+                          // the headers are riddled with unused parameters and emit
+                          // hundreds of warnings: suppress them.
+                          "-Wno-unused-parameter",
+                          "-Wno-sign-compare",
+                          "-Wno-missing-field-initializers",
```
Comment (anxiousmodernman) [inline build.rs]: 👍

File: extract-features.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
+#!/usr/bin/python
```
Comment (anxiousmodernman) [inline extract-features.py:1]: Is the idea here that one would do something like this?

```
$ python3 extract-features.py >> Cargo.toml
```

**Review by anxiousmodernman — COMMENTED** [review]
The build.rs changes that recursively add source files and headers looks nice to me.

I am actually unclear on the consequences of the `[features]` changes. Are there any alternatives  to configuring them like this?

File: extract-features.py (hunk 3 above)
Comment (wez) [reply extract-features.py:1]: Yeah, that's exactly what I did here.  Updating/evolving it over time is a bit manual and clunky

**Comment by wez** [comment]
re: the features question, I don't really know of anything better, but I'm still relatively new to rust.  It does feel like there are probably more of them included here than most folks will need.

**Comment by anxiousmodernman** [comment]
Some of the errors I encountered earlier were related to (breaking) changes in generated code (see #6 ). We've been chasing bindgen a little bit, but given how important that project is, it feels right to stay current. That work is on [this branch](https://github.com/jamesmunns/nrf52dk-sys/commits/6_update_bindgen), and anyone who uses those referenced constants will be broken, so if it's merged we'll need a minor number bump.

Over the next few days I will:
* look for a more cargo-ish way to do what the python script does; this script looks useful but I'd like to avoid checking it in
* learn to grok the features stuff
* rebase my branch off this PR to test against latest bindgen on linux
* downgrade to test against old bindgen on linux

Then I might come back with a few more requests on this around making the adafruit toolchain paths configurable, etc. I'm **colemanx** in IRC so if you ever see me feel free to hit me up.

**Comment by jamesmunns** [comment]
Hey @anxiousmodernman, thanks for following up with this. I'm out of the country for the next week with bad internet, so feel free to merge when you (and @wez) feel comfortable, and `@` me or email me (its on my profile) if you have any direct questions.

I am okay with semver bumps, and feel free to mandate a minimum (or specific) version of Bindgen, just make sure to update the install guides (just grep the repo for the current version to make sure you catch it everywhere).

File: extract-features.py (hunk 3 above)
Comment (anxiousmodernman) [reply extract-features.py:1]: We may be able to use `println!` in build.rs to build the set of features dynamically http://doc.crates.io/build-script.html See `rustc-cfg=FEATURE`

**Comment by anxiousmodernman** [comment]
@wez Changes to generated code introduced some conflicts here. I've rebased those changes in the `wez_pr` branch, but I haven't gotten it to build yet. I'll keep trying.

**Pull request closed without merging**